src/hotspot/share/classfile/stringTable.cpp

```diff
@@ -323,80 +374,137 @@ void StringTable::update_needs_rehash(bool rehash) {
+const jchar *StringTable::to_unicode(StringWrapper wrapped_str, int &len, TRAPS) {
+  switch (wrapped_str.type) {
+  case StringType::UnicodeStr:
+    return wrapped_str.unicode_str;
+  case StringType::OopStr:
+    return java_lang_String::as_unicode_string(wrapped_str.oop_str(), len, CHECK_NULL);
+  case StringType::SymbolStr: {
+    const char* utf8_str = wrapped_str.symbol_str->get_utf8();
+    int unicode_length = UTF8::unicode_length(utf8_str, wrapped_str.symbol_str->utf8_length());
+    jchar *chars = NEW_RESOURCE_ARRAY(jchar, unicode_length);
+    UTF8::convert_to_unicode(utf8_str, chars, unicode_length);
+    len = unicode_length;
+    return chars;
+  }
+  case StringType::UTF8Str: {
+    int unicode_length = UTF8::unicode_length(wrapped_str.utf8_str);
+    jchar *chars = NEW_RESOURCE_ARRAY(jchar, unicode_length);
+    UTF8::convert_to_unicode(wrapped_str.utf8_str, chars, unicode_length);
+    len = unicode_length;
+    return chars;
+  }
+  }
+  ShouldNotReachHere();
+  return nullptr;
+}
+
+Handle StringTable::to_handle(StringWrapper wrapped_str, int len, TRAPS) {
+  switch (wrapped_str.type) {
+  case StringType::OopStr:
+    return wrapped_str.oop_str;
+  case StringType::UnicodeStr:
+    return java_lang_String::create_from_unicode(wrapped_str.unicode_str, len, THREAD);
+  case StringType::SymbolStr:
+    return java_lang_String::create_from_symbol(wrapped_str.symbol_str, THREAD);
+  case StringType::UTF8Str:
+    return java_lang_String::create_from_str(wrapped_str.utf8_str, THREAD);
+  }
+  ShouldNotReachHere();
+  return Handle();
+}
+
 // Interning
 oop StringTable::intern(Symbol* symbol, TRAPS) {
   if (symbol == nullptr) return nullptr;
-  ResourceMark rm(THREAD);
-  int length;
-  jchar* chars = symbol->as_unicode(length);
-  Handle string;
-  oop result = intern(string, chars, length, CHECK_NULL);
+  int length = symbol->utf8_length();
+  StringWrapper name(symbol);
+  oop result = intern(name, length, CHECK_NULL);
   return result;
 }
 
 oop StringTable::intern(oop string, TRAPS) {
   if (string == nullptr) return nullptr;
-  ResourceMark rm(THREAD);
-  int length;
+  int length = java_lang_String::length(string);
   Handle h_string (THREAD, string);
-  jchar* chars = java_lang_String::as_unicode_string(string, length,
-                                                     CHECK_NULL);
-  oop result = intern(h_string, chars, length, CHECK_NULL);
+  StringWrapper name(h_string);
+  oop result = intern(name, length, CHECK_NULL);
   return result;
 }
 
 oop StringTable::intern(const char* utf8_string, TRAPS) {
   if (utf8_string == nullptr) return nullptr;
-  ResourceMark rm(THREAD);
-  int length = UTF8::unicode_length(utf8_string);
-  jchar* chars = NEW_RESOURCE_ARRAY(jchar, length);
-  UTF8::convert_to_unicode(utf8_string, chars, length);
-  Handle string;
-  oop result = intern(string, chars, length, CHECK_NULL);
+  int length = static_cast<int>(strlen(utf8_string));
```

Comment:
Is the source for such a string always a constant pool entry, and so known to < 64K in length?

Reply:
I am not sure. The previous implementation stored length as `int`, so my assumption is that it would be fine. If it is a problem, I could take some time to change the various lengths to `size_t`s instead.

Reply:
See the preamble in utf8.hpp. The unicode length of a utf8 string is assumed/required to fit in an int, but the raw utf8 length need not. Changing to use size_t may have a significant fanout however.